jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt

```diff
@@ -0,0 +1,103 @@
+
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.ToggleAction
+import com.intellij.openapi.actionSystem.ex.ComboBoxAction
+import software.amazon.awssdk.services.toolkittelemetry.model.Unit as MetricUnit
+import software.aws.toolkits.core.telemetry.TelemetryNamespace
+import software.aws.toolkits.jetbrains.services.telemetry.TelemetryService
+import javax.swing.Icon
+
+object ToolkitActionPlaces {
+    val EXPLORER_WINDOW = "ExplorerToolWindow"
+}
+
+// Constructor signatures:
+//  public AnAction(){
+//  }
+//  public AnAction(Icon icon){
+//    this(null, null, icon);
+//  }
+//  public AnAction(@Nullable String text) {
+//    this(text, null, null);
+//  }
+//  public AnAction(@Nullable String text, @Nullable String description, @Nullable Icon icon) {
+//    <logic>
+//  }
+abstract class AnActionWrapper : TelemetryNamespace, AnAction {
+    constructor(): super()
+    constructor(text: String? = null, description: String? = null, icon: Icon? = null):
+        super(text, description, icon)
+
+    /**
+     * Consumers should use doActionPerformed(e: AnActionEvent)
+     */
+    final override fun actionPerformed(e: AnActionEvent) {
+        doActionPerformed(e)
+        telemetry.record(getNamespace()) {
+            datum(e.place) {
+                value(1.0)
+                unit(MetricUnit.COUNT)
+            }
+        }
+    }
+
+    abstract fun doActionPerformed(e: AnActionEvent)
+
+    companion object {
+        val telemetry = TelemetryService.getInstance()
+    }
+}
+
+abstract class ComboBoxActionWrapper : TelemetryNamespace, ComboBoxAction() {
+    /**
+     * Consumers should use doActionPerformed(e: AnActionEvent)
+     */
+    final override fun actionPerformed(e: AnActionEvent) {
+        doActionPerformed(e)
+        telemetry.record(getNamespace()) {
+            datum(e.place) {
+                value(1.0)
+                unit(MetricUnit.COUNT)
+            }
+        }
+    }
+
+    open fun doActionPerformed(e: AnActionEvent) = super.actionPerformed(e)
+
+    companion object {
+        val telemetry = TelemetryService.getInstance()
+    }
+}
+
+abstract class ToogleActionWrapper : TelemetryNamespace, ToggleAction {
+    constructor(): super()
+    constructor(text: String? = null, description: String? = null, icon: Icon? = null):
+        super(text, description, icon)
+
+    // this will be repeatedly called by the IDE, so we likely do not want telemetry on this,
+    // but keeping this to maintain API consistency
+    final override fun isSelected(e: AnActionEvent): Boolean = doIsSelected(e)
+
+    final override fun setSelected(e: AnActionEvent, state: Boolean) {
+        doSetSelected(e, state)
+        telemetry.record(getNamespace()) {
+            datum(e.place) {
+                value(1.0)
+                unit(MetricUnit.COUNT)
+            }
+        }
+    }
```

Comment:
Can you wrap the report of count metrics to an extension method? There is no point to repeat three times for the same thing.

Reply:
updated